tests/test_publish_draft_services.py

```diff
@@ -0,0 +1,271 @@
+                # if it's a dry_run then the script never discovers that it's already published so will call
+                # either way
+            ) if dry_run or not skip_docs_if_published else None,
+            mock.call(
+                ExactIdentity(mock_data_api_client),
+                "g-cloud-123",
+                DraftServiceStub(**draft_service_kwargs_by_id[492004]).response(),
+                AnyStringMatching(r"555.*") if dry_run else "7777492004",
+                dry_run
+            ),
+            mock.call(
+                ExactIdentity(mock_data_api_client),
+                "g-cloud-123",
+                DraftServiceStub(**draft_service_kwargs_by_id[492044]).response(),
+                "4448882220",
+                dry_run
+            ) if not skip_docs_if_published else None,
+            mock.call(
+                ExactIdentity(mock_data_api_client),
+                "g-cloud-123",
+                DraftServiceStub(**draft_service_kwargs_by_id[492444]).response(),
+                AnyStringMatching(r"555.*") if dry_run else "7777492444",
+                dry_run
+            ),
+            mock.call(
+                ExactIdentity(mock_data_api_client),
+                "g-cloud-123",
+                DraftServiceStub(**draft_service_kwargs_by_id[497005]).response(),
+                AnyStringMatching(r"555.*") if dry_run else "7777497005",
+                dry_run
+            ),
+        ) if x is not None
+    ]
+
+
+@pytest.mark.parametrize("dry_run", (False, True,))
+def test_copy_draft_documents(dry_run):
+    mock_draft_bucket = mock.create_autospec(S3, bucket_name="ducky")
+    mock_draft_bucket.path_exists.return_value = True
+
+    mock_documents_bucket = mock.create_autospec(S3, bucket_name="puddeny-pie")
+
+    def mock_documents_bucket_copy_side_effect(src_bucket, src_key, target_key, **kwargs):
+        # let's say somehow this already exists
+        if target_key == "g-cloud-123/documents/171400/2229090909-kidney.pdf":
+            raise ValueError("Target key already exists in S3.")
+        return mock.Mock()
+    mock_documents_bucket.copy.side_effect = mock_documents_bucket_copy_side_effect
+
+    mock_data_api_client = mock.create_autospec(DataAPIClient)
+
+    copy_draft_documents(
+        mock_draft_bucket,
+        mock_documents_bucket,
+        (
+            "chinChopperURL",
+            "beeoTeeTomURL",
+            "madcapURL",
+            "boldAsBrassURL",
+        ),
+        "https://boyish.gam/bols/",
+        mock_data_api_client,
+        "g-cloud-123",
+        DraftServiceStub(
+            id=379001,
+            framework_slug="g-cloud-123",
+            supplier_id="171400",
+            service_id="2229090909",
+            **{
+                "chinChopperURL": "https://gin.ger.br/ead/g-cloud-123/submissions/171400/379001-steak.pdf",
+                "beeoTeeTomURL": "https://gin.ger.br/ead/g-cloud-123/submissions/171400/379001-kidney.pdf",
```

Comment:
So this one should raise the ValueError, correct? Do we assert that?

Reply:
Ooh well spotted - I realized this one was a tricky situation because the only side effect it has is a possible log message going out. _But_ the thing we really care about the script's behaviour in handling this exception is that it doesn't cause it to abort the run. So in essence the fact that the script continues and checks the following drafts _is_ the assertion.

Reply:
Ah I see - we raise if it's anything other than 'file is already here'. Cool 👍